### DISCUSSION TITLE
How to toggle a bool value of an atom with only 1 function?

### DISCUSSION THREAD

**Discussion by kenildev** [pmndrs/jotai#1559] [Q&A]
So I have a showMenu bool
![q1](https://user-images.githubusercontent.com/117978788/201392351-81a11a49-2703-4f25-996b-d1078f2652ca.png)

And then in a button onClick I can simply do: 

![q2](https://user-images.githubusercontent.com/117978788/201392663-64844701-e72b-4a81-b9d2-8d02cd90f156.png)

My question is if I want to toggle the atom value in another place do I have to do it all over again??

Do I have to once again useAtom and all of that or is there some way to do this in Jotai? I want a function which toggles the showMenu bool and once I want to use I can simply import it and then call it?

![image](https://user-images.githubusercontent.com/117978788/201392979-50248dbb-a1b9-4bb2-925d-8adda473fbc2.png)

and then import it: 
![image](https://user-images.githubusercontent.com/117978788/201393065-fa808ed4-49c6-4a4b-9850-411d34168f8e.png)

and that will shorten my code.

However when I do that I get this error: 

![jotai js question](https://user-images.githubusercontent.com/117978788/201393243-79430098-6017-4b0a-844b-405284f980cd.png)

Is there some way to imlement this because calling useAtom and all of that again seems very tedious and I could do that with globalContext + useReducer but here I cannot. 

Thanks

**Comment by edongashi** [comment]
The setter accepts a function of `prevState => nextState`. We can pass it a simple `not` function:

```js
function MyComponent() {
  const setShowMenu = useSetAtom(showM);
  
  return (
    <button onClick={() => setShowMenu(showMenu => !showMenu)}>Toggle menu</button>
  );
}
```

^ This prevents unnecessary re-renders because we don't subscribe to the current value.

If you don't want the details leaking out, you can write a wrapper hook that returns the action:

```js
function useToggleShowMenu() {
  const setShowMenu = useSetAtom(showM);
  return useCallback(() => {
    setShowMenu(showMenu => !showMenu);
  }, [setShowMenu]);
}

// ...

function MyComponent() {
  const toggleShowMenu = useToggleShowMenu();
 
  return (
    <button onClick={toggleShowMenu}>Toggle menu</button>
  );
}
```

> However when I do that I get this error:

This happens because `onClick` runs outside of the render cycle. Hooks can only be written in the component function and unconditionally.

**Comment by dai-shi** [comment]
Check out atomWithReducer.

**Reply by QzCurious** [reply]
Hi @dai-shi.

The doc is not listing `atomWithReducer` (meanwhile, I can reach it by search).

![image](https://user-images.githubusercontent.com/38932402/236110230-7d59f331-985a-44e8-bdf7-51d5b1fecc65.png)

I had no experience with gatsby. But I guess you're [filtering out](https://github.com/pmndrs/jotai/blob/main/website/src/components/docs.js#L12) `atomWithReducer` (which is marked as unpublished) from nav.

![image](https://user-images.githubusercontent.com/38932402/236110438-81c8d75b-2c71-4722-a8b9-73d9bcc7e607.png)

Could you publish `atomWithReducer` doc?

**Reply by dai-shi** [reply]
It's intentional, but maybe there's some ways to mitigate it? Like "Others" section under "Utilities"? @sandren 
(Also internal `published` flag seems bad naming as people assumes it might be unofficial/unsupported?)

**Reply by QzCurious** [reply]
Yes. At this point, I still not sure if `atomWithReducer` is supported. And hesitant to make a feature request for `atomWithReducerWithDefault`.

**Reply by dai-shi** [reply]
`atomWithReducer` is supported.
`atomWithReducerWithDefault` would not be provided, but it might be good for [atom creators recipe](https://jotai.org/docs/recipes/atom-creators) or a third-party package.

**Reply by sandren** [reply]
> It's intentional, but maybe there's some ways to mitigate it? Like "Others" section under "Utilities"? @sandren (Also internal `published` flag seems bad naming as people assumes it might be unofficial/unsupported?)

Yes, let's rename the flag to something more explicit.
